### PR TITLE
Refactor & move get_robots_meta()

### PR DESCRIPTION
**Pull request by arnaudbroes**
Issue #2684

## Proposed changes

Refactors the robots meta tag code even further and moves it to its own dedicated class file.

This refactors it up to the point that get_robots_meta() can be read and understood by literally anyone.

## Types of changes

- Improves existing code

## Checklist

- [x] I've selected the appropriate branch (ie x.y rather than master).
- [x] I've created an appropriate title, descriptive of what the PR does.
- [x] Travis-ci runs with no errors.

## Testing instructions

Test all that is in the release testing spreadsheet with regard to the robots meta tag.

## Further comments

I was planning on pushing this together with the unit tests once I'm done with those, but since the 3.3 release is nearing I'm going ahead and pushing this first.

### CODE REVIEW

**Comment by michaeltorbert** [comment]
Please post a link to the release testing spreadsheet.

**Comment by arnaudbroes** [comment]
https://docs.google.com/spreadsheets/d/1RMeC_LDZd9jzbcuKmCadRYOR4xvVJy097FP01hyloO8/edit?usp=sharing

**Comment by michaeltorbert** [comment]
Noindex for taxonomy term pages is broken.

**Comment by arnaudbroes** [comment]
> Noindex for taxonomy term pages is broken.

This works fine on my end. Do you have steps to reproduce?

image

**Comment by michaeltorbert** [comment]
When doing it from the General Settings page it doesn't work anymore.

**Comment by arnaudbroes** [comment]
> When doing it from the General Settings page it doesn't work anymore.

So I created a CPT with a custom taxonomy, created a term and assigned the CPT to it and then checked the box next to "Use noindex for Taxonomy Archives" in the General Settings and it seems to be working. Might be worth doing a call to discuss this.

**Comment by arnaudbroes** [comment]
@michaeltorbert bug has been fixed and PR can be retested.

**Comment by wpsmort** [comment]
@arnaudbroes I tested this and found that two warnings are output on any 404 page:

First parameter must either be an object or the name of an existing class in all-in-one-seo-pack/inc/general/aioseop-robots-meta.php on line 128

First parameter must either be an object or the name of an existing class in all-in-one-seo-pack/inc/general/aioseop-robots-meta.php on line 132

This happens regardless of what's checked under Noindex Settings.

To reproduce this, install the code and go to a URL that returns a 404 page and you should see the warnings.

**Comment by wpsmort** [comment]
This is tested and is good.

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (EkoJR) [inline .vscode/launch.json]: This file is for visual studio. Was this file accidently added?

File: inc/general/aioseop-robots-meta.php [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+			$post_type = 'page';
+		}
+
+		if ( $this->has_post_meta() ) {
+			$post_meta_noindex  = $this->get_meta_value( 'noindex' );
+			$post_meta_nofollow = $this->get_meta_value( 'nofollow' );
+		}
+
+		if ( 'on' === $post_meta_noindex || $this->is_noindexed_paginated_page( $page_number ) ||
+		$this->is_noindexed_password_protected_post() || $this->is_noindexed_tax() || $this->is_noindexed_singular( $post_type, $post_meta_noindex ) ) {
```
Comment (EkoJR) [inline inc/general/aioseop-robots-meta.php]: This is a bit odd of a style.

File: inc/general/aioseop-robots-meta.php [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+	 * Checks whether the current page is paginated and should be noindexed.
+	 *
+	 * @since 3.3.0
+	 *
+	 * @param int $page_number
+	 * @return bool
+	 */
+	private function is_noindexed_paginated_page( $page_number ) {
+		global $aioseop_options;
+		if ( ! empty( $aioseop_options['aiosp_paginated_noindex'] ) && $page_number > 1 ) {
```
Comment (EkoJR) [inline inc/general/aioseop-robots-meta.php]: Yoda missing

File: inc/general/aioseop-robots-meta.php [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+	 * Checks whether the current page is paginated and should be nofollowed.
+	 *
+	 * @since 3.3.0
+	 *
+	 * @param int $page_number
+	 * @return bool
+	 */
+	private function is_nofollowed_paginated_page( $page_number ) {
+		global $aioseop_options;
+		if ( ! empty( $aioseop_options['aiosp_paginated_nofollow'] ) && $page_number > 1 ) {
```
Comment (EkoJR) [inline inc/general/aioseop-robots-meta.php]: Yoda missing

File: inc/general/aioseop-robots-meta.php [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+	 *
+	 * @param string $post_type
+	 * @param string $post_meta_noindex
+	 *
+	 * @return bool
+	 */
+	private function is_noindexed_singular( $post_type, $post_meta_noindex ) {
+		global $aioseop_options;
+		if ( is_singular() && '' === $post_meta_noindex &&
+		! empty( $aioseop_options['aiosp_cpostnoindex'] ) && in_array( $post_type, $aioseop_options['aiosp_cpostnoindex'] ) ) {
```
Comment (EkoJR) [inline inc/general/aioseop-robots-meta.php]: This is also a bit odd for indenting.

File: inc/general/aioseop-robots-meta.php [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+	 *
+	 * @param string $post_type
+	 * @param string $post_meta_follow
+	 *
+	 * @return bool
+	 */
+	private function is_nofollowed_singular( $post_type, $post_meta_follow ) {
+		global $aioseop_options;
+		if ( is_singular() && '' === $post_meta_follow &&
+		! empty( $aioseop_options['aiosp_cpostnofollow'] ) && in_array( $post_type, $aioseop_options['aiosp_cpostnofollow'] ) ) {
```
Comment (EkoJR) [inline inc/general/aioseop-robots-meta.php]: This is also another IF statement that has an odd coding style.

File: modules/aioseop_opengraph.php [hunk 7]
```diff
@@ -1048,7 +1048,7 @@ function add_meta() {
 
 			if ( ! empty( $aioseop_options['aiosp_hide_paginated_descriptions'] ) ) {
 				$first_page = false;
-				if ( $aiosp->get_page_number() < 2 ) {
+				if ( aioseop_get_page_number() < 2 ) {
```
Comment (EkoJR) [inline modules/aioseop_opengraph.php]: Yoda missing

File: inc/general/aioseop-robots-meta.php (hunk 3 above)
Comment (arnaudbroes) [reply inc/general/aioseop-robots-meta.php]: Resolved.

**Pull request merged**